### PR TITLE
Support custom MonoCut attrs (with special support for loading Arrays)

### PR DESCRIPTION
**Pull request by pzelasko**
This seems to work in trivial scenarios, but needs more testing. 

See the unit test for an example of usage: storing some data in HDF5, assigning its manfiest to `ivector`, and then calling `cut.load_ivector()` without Cut having that method defined anywhere.

This is fairly painless, but needs more testing especially with `TemporalArray` to see if padding and truncation are behaving correctly. `MixedCut` is not going to have this functionality (it's not well defined what it means to e.g. mix embeddings...)

For discussion/design/motivation see #453 and #393

### CODE REVIEW

**Comment by pzelasko** [comment]
It also needs testing for serialization (but I think it should work out of the box)

File: lhotse/cut.py [hunk 1]
```diff
@@ -760,6 +760,50 @@ class MonoCut(Cut):
...
+        If ``item`` is found in ``custom``, we'll return ``custom[item]``.
+        If ``item`` starts with "load_", we'll assume the name of the relevant
+        attribute comes after that, and that value of that field is of type
+        Array or TemporalArray. We'll return it's ``load`` method to call by the user.
+        """
+        from lhotse.array import Array, TemporalArray
+
+        custom = self.custom
+        if custom is None:
+            raise AttributeError()
```
Comment (csukuangfj) [inline lhotse/cut.py:788]: Shall we also display the name of the accessed attribute in the exception?
Just for better diagnostics.

File: test/cut/test_custom_attrs.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import numpy as np
+
+from lhotse import LilcomHdf5Writer, MonoCut
+from lhotse.array import Array
+
+
+def test_cut_load_array():
+    array = np.arange(20).astype(np.float32)
+    with NamedTemporaryFile(suffix=".h5") as f, LilcomHdf5Writer(f.name) as writer:
+        manifest = Array.store(key="utt1", value=array, writer=writer)
```
Comment (danpovey) [inline test/cut/test_custom_attrs.py:12]: I was a little confused here by the name "manifest" as I had assumed "manifest" always refers to a collection of metadata, but I suppose it refers to an item of metadata here?  I wonder if those should be given different names...

Comment (pzelasko) [reply test/cut/test_custom_attrs.py:12]: This is correct — calling Array.store returns a manifest for the stored data. Not sure how to balance informativeness and verbosity — Array.store_and_create_manifest is a mouthful. Do you have any suggestions?

Comment (pzelasko) [reply test/cut/test_custom_attrs.py]: I changed it to `manifest = writer.store_array(...)` in the other PR, maybe this is more helpful?

File: test/cut/test_custom_attrs.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        )
+        cut_trunc = cut.truncate(duration=5.0)
+
+        alignment_piece = cut_trunc.load_alignment()
+        assert alignment_piece.shape == (13,)  # 5.0 / 0.4 == 12.5 ~= 13
+        np.testing.assert_equal(alignment[:13], alignment_piece)
+
+
+@pytest.mark.parametrize("pad_value", [-1, 0])
+def test_cut_load_temporal_array_pad(pad_value):
```
Comment (pzelasko) [inline test/cut/test_custom_attrs.py:86]: This test shows how arrays read via padded cuts are auto-padded.

**Comment by pzelasko** [comment]
The auto-padding code seems to work fine including edge-cases; I will start working on the next PR that shows how to use this in a PyTorch Dataset. I might do some refactoring too to make the Cut/MixedCut `__setattr__` code easier to comprehend/maintain.

File: test/cut/test_custom_attrs_randomized.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            ),
+            label=f"Padded cut duration",
+        )
+        padded = cut.pad(
+            duration=duration,
+            direction=pad_direction,
+            pad_value_dict={"codebook_indices": -1},
+        )
+        # Test the invariants
+        array = padded.load_codebook_indices()
```
Comment (pzelasko) [inline test/cut/test_custom_attrs_randomized.py:55]: This randomized test runs 500 times with cuts of different sampling_rates, durations, etc. It attaches a 1D TemporalArray called "codebook_indices" with random ints, pads the cut, and reads out the padded array from the padded cut.

File: lhotse/testing/fixtures.py [hunk 5]
```diff
@@ -124,3 +130,13 @@ def _with_alignment(
...
+
+    def _with_custom_temporal_array(self, cut: MonoCut, frame_shift: Seconds) -> None:
+        d = TemporaryDirectory()
+        self.dirs.append(d)
+        num_frames = seconds_to_frames(cut.duration, frame_shift=frame_shift)
+        array = np.random.randint(256, size=(num_frames,))
+        with NumpyHdf5Writer(d.name) as storage:
+            cut.codebook_indices = storage.store_array(
+                key="ali1", value=array, frame_shift=frame_shift, temporal_dim=0
+            )
```
Comment (pzelasko) [inline lhotse/testing/fixtures.py:142]: This snippet shows how "codebook_indices" are stored and attached to a Cut manifest.

**Comment by danpovey** [comment]
Thanks!  @glynpu is working on using this framework to dump encodings of embeddings for Librispeech, and train on them as an auxiliary loss.  Then we'll go on to try to handle the multi-dataset case, with Gigaspeech.

**Comment by pzelasko** [comment]
Great. Keep me updated on the progress and if you need anything else. I’ll write some docs and examples for this mechanism soon.

**Comment by danpovey** [comment]
Thanks!
